### ISSUE TITLE
in .AuditAction can't escape insert into db

### ISSUE THREAD

**Issue by mhamri** [thepirat000/Audit.NET#122]
let say i have some condition that i mentioned in #121 i want to escape an entity to insert into the db. 

```C#
Audit.Core.Configuration.Setup()
    .UseEntityFramework(ef => ef
        .AuditTypeExplicitMapper(m => m
            .Map<Order, Audit_Order>()
            .Map<OrderItem, Audit_OrderItem>()
            .AuditEntityAction<IAudit>((evt, entry, auditEntity) =>
            {
                // here i understand this entity shouldn't be recorded, 
                // i want to stop the audit entity to be recorded, but i can't return anything that stop the auditEntity from saving
                // the best solution is auditEntity=null;
                // then auditaction should check if this is returned as null or not. 
                // atm it's not checking it. so it insert null value into the db!
            })
        )
    );
```

so my suggestion is to put a checking at line 71 in `~\Audit.NET\src\Audit.EntityFramework\Providers\EntityFrameworkDataProvider.cs`

```C#
                    if (mappedType != null)
                    {
                        var auditEntity = CreateAuditEntity(type, mappedType, entry);
                        _auditEntityAction?.Invoke(efEvent, entry, auditEntity);
                        if (auditEntity==null) continue; //after invoke check if it returned null or not
#if NET45
                        dbContext.Set(mappedType).Add(auditEntity);
#else
                        dbContext.Add(auditEntity);
#endif
                        save = true;
                    }
```

**Comment by thepirat000** [comment]
We'll need to think a different way to do this. What you suggest is not possible unless we mark the auditEntity parameter as `ref` and that will be a breaking change.

**Comment by mhamri** [comment]
why `ref`? a class is refrence type? by default it should works

**Comment by thepirat000** [comment]
```
var action = new Action<object>(obj =>
{
	obj = null;
});

var myObj = new List<int>() { 1,2,3 };
action.Invoke(myObj);
Console.WriteLine(myObj);
```

So you say code above prints null?

**Comment by thepirat000** [comment]
I think the most flexible way is to optionally allow passing a Func that returns bool instead of the Action. 

I will work on this

**Comment by thepirat000** [comment]
This was fixed on version 12.0.2, please update and re-test.

Now you can return a _false_ boolean value on the `AuditEntityAction` to avoid that entity to being logged on the audit output.

Refer to the updated documentation [here](https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.EntityFramework/README.md#ef-provider-configuration). 

Feel free to re-open if you're still having problems

**Comment by mhamri** [comment]
@thepirat000 i didn't know about delegate is passed as a copy of values but still can call the child object. it's strange. thanks for great lesson. yup it is a good idea

**Comment by axceligent** [comment]
I updated to latest version and I am returning false on occasion that I don't want to save with the following code :
```C#
Audit.Core.Configuration.Setup()
.UseEntityFramework(ef => ef
    .AuditTypeMapper(type =>
    {
        if (!typeof(IAuditEnabler).IsAssignableFrom(type))
        {
            return null;
        }

        var intermediateType = Type.GetType($"{type.Name}_Audit");
        if (intermediateType != null && typeof(IAuditEnabler).IsAssignableFrom(intermediateType))
        {
            return intermediateType;
        }

        return typeof(AuditLog);
    })
    .AuditEntityAction((auditEvent, auditEntry, changedOn) =>
    {
        var changeObserved = false;
        switch (auditEntry.Action)
        {
            case "Update":

                foreach (var auditEntryChange in auditEntry.Changes)
                {
                    if (auditEntryChange.NewValue.Equals(auditEntryChange.OriginalValue))
                    {
                        changeObserved = true;
                    }
                }
                break;

            default:
                changeObserved = true;
                break;
        }
        return changeObserved;
    })

);
```

but the audit.net still saving the empty line into the db
![image](https://user-images.githubusercontent.com/39044043/39984596-4ba08226-578e-11e8-86ee-c963a353b911.png)

also i have one extra question, is it fine to put `Audit.Core.Configuration.Setup()` in the `Startup.cs` or should be placed in the `DbContext` ? in general, what is the right place to put this configuration?

**Comment by axceligent** [comment]
i just noticed the change is not persistant for line https://github.com/thepirat000/Audit.NET/blob/31257389c30d949707087ddc23d55da9f6f18f9d/src/Audit.EntityFramework/Providers/EntityFrameworkDataProvider.cs#L124
like the `InsertEvent`. returning false doesn't affect insert into the db

**Comment by thepirat000** [comment]
Sorry about that... I forget to make the change on the SaveAsync method and also forget to add the Async version of the unit tests... 

This is fixed now on version 12.1.1

Feel free to re-open this if you're still having problems.
